cmd/registry/core/references.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sort"
+	"strings"
+
+	"github.com/apigee/registry/rpc"
+	openapi_v2 "github.com/googleapis/gnostic/openapiv2"
+	openapi_v3 "github.com/googleapis/gnostic/openapiv3"
+	protoparser "github.com/yoheimuta/go-protoparser/v4"
+	"github.com/yoheimuta/go-protoparser/v4/parser"
+)
+
+// NewReferencesFromOpenAPIv2 ...
+func NewReferencesFromOpenAPIv2(document *openapi_v2.Document) (*rpc.References, error) {
+	return nil, nil
+}
+
+// NewReferencesFromOpenAPIv3 ...
+func NewReferencesFromOpenAPIv3(document *openapi_v3.Document) (*rpc.References, error) {
+	return nil, nil
+}
+
+// NewReferencesFromZippedProtos ...
+func NewReferencesFromZippedProtos(b []byte) (*rpc.References, error) {
+	// create a tmp directory
+	dname, err := ioutil.TempDir("", "registry-protos-")
+	if err != nil {
+		return nil, err
+	}
+	// whenever we finish, delete the tmp directory
+	defer os.RemoveAll(dname)
+	// unzip the protos to the temp directory
+	_, err = UnzipArchiveToPath(b, dname)
+	if err != nil {
+		return nil, err
+	}
+	// process the directory
+	internals, externals, err := internalsAndExternalsForPath(dname)
+	externals = removeInternalsAndDuplicatesFromExternals(internals, externals)
+	sort.Strings(internals)
+	sort.Strings(externals)
```

Comment:
Nit: Consider changing these methods to `referencesForPath(path string) (*rpc.References, error)`, `deduplicate(refs *rpc.References)`, `filterExternals(refs *rpc.References, exclusions []string)`, and `sort(refs *rpc.References)`. A slightly heavier but perhaps cleaner approach would be to have a separate type that could have methods for these and one to convert to RPC.

The current directory processing code makes sense but I think if we modify it or reuse these functions in the future there are some bugs that could be easily introduced, e.g. forgetting the "remove" helper doesn't mutate the slice.

Reply:
OK - I'm happy to do the renaming (is there a style guide reference that could help me get this right in the future?) but don't want to overbuild this because I think it is still somewhat experimental.

Reply:
There isn't any specific style guide for function naming. My suggestion comes from the single responsibility principle, having separate functions `foo()` and `bar()` instead of `doFooAndBar()`. My name choices were best effort from my reading of your code, but could be different.